More verbose error message when login nonce check fails

peterhashair opened the pull request:
### Description:

Fixes: #16607 
split nonce verifies into 3 parts. By extending the verifyNonce function pass 4th params with true or false, the default won't show details.  I am not sure about the security policy around this one, I know some errors required generic error messages to prevent getting too many details, but for this case, I don't think will applied. 

-   Token mismatch (InvalidNonceToken)
-   Referrer (InvalidNonceReferrer)
-   Origin (InvalidNonceOrigin)

- deprecated getMessageExceptionNoAccess Login/controller
- deprecated InvalidNonceOrHeadersOrReferrer in lang
- move getMessageExceptionNoAccessWhenInsecureConnectionMayBeUsed from Login/controller to core/Nonce
- fix missing 2 variables in translation.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

Commit:
add http to https error message on nonce check

add http to https error message on nonce check

Findus23 commented:
Honestly I meant in #16607 that we should log or output more precisely which one of the many checks in https://github.com/matomo-org/matomo/blob/679e73f1236969db0c2d767655cb84456a727d24/core/Nonce.php#L70 and https://github.com/matomo-org/matomo/blob/06d43857c48ada2fa7f1ad18a8309e8826c0e413/core/Url.php#L547 fails instead of leaving the user guessing and debugging Matomo code. 

Adding the sentence by @Starker3 is also fine, but doesn't affect the core issue I meant.

@peterhashair unrelated, but it would be great if you could link issues in the PR description, so one can directly jump back to them instead of searching for the linked issue in the sidebar. And if you use something like `fixes #1234`, you don't need to manually link them either.

Commit:
Merge branch '4.x-dev' into m-16607-error-message-on-nonce-fails

peterhashair commented:
@Findus23 add the fixes number, I don't even know that, thanks for the hints :). If I understand correctly, we want to split the verifyNonce into details with their own error messages, instead of combining them into one error log?

Commit:
add todo

add todo

Commit:
Merge branch '4.x-dev' into m-16607-error-message-on-nonce-fails

Commit:
update Nonce on details

update Nonce on details

Commit:
Merge branch '4.x-dev' into m-16607-error-message-on-nonce-fails

tsteur commented on core/Nonce.php:
```diff
@@ -65,26 +65,27 @@ public static function getNonce($id, $ttl = 600)
      * @return bool `true` if valid; `false` otherwise.
      */
-    public static function verifyNonce($id, $cnonce, $expectedReferrerHost = null)
+    public static function verifyNonce($id, $cnonce, $expectedReferrerHost = null, bool $withErrorMessage = false)
```
@peterhashair could we maybe rather have two different methods `verifyNonceWithErrorMessage` (there's likely better wording for it) and `verifyNonce`? That will reduce the complexity of the method a bit and often boolean parameters indicate that it should be maybe two methods. You can still reuse the code like 


```
public static function verifyNonce($id, $cnonce, $expectedReferrerHost = null){
    $error = self::verifyNonceWith...($id, $cnonce, $expectedReferrerHost);
    return empty($error);
}
```

Commit:
Merge branch '4.x-dev' into m-16607-error-message-on-nonce-fails

Commit:
update verify function

update verify function

Commit:
Merge branch '4.x-dev' into m-16607-error-message-on-nonce-fails

tsteur commented on core/Nonce.php:
```diff
@@ -65,38 +65,59 @@ public static function getNonce($id, $ttl = 600)
         $referrer = Url::getReferrer();
         if (empty($expectedReferrerHost) && !empty($referrer) && !Url::isLocalUrl($referrer)) {
-            return false;
+            return Piwik::translate('Login_InvalidNonceReferrer');
```
@peterhashair is there any chance we can slightly tweak the two same `InvalidNonceReferrer` messages to have them bit different? when users report us that they have a problem we'll then better know what the problem is.

Commit:
Merge branch '4.x-dev' into m-16607-error-message-on-nonce-fails

Commit:
update referral message on different expect host

update referral message on different expect host

Commit:
update unexpected Referrer head

update unexpected Referrer head

peterhashair commented:
@tsteur split the error message,  but I don't think $expectedReferrerHost is used anywhere.  Should we set a variable in config.ini.php that reflected that value?

tsteur commented on plugins/Login/lang/en.json:
```diff
@@ -5,7 +5,11 @@
         "ExceptionPasswordMD5HashExpected": "The password parameter is expected to be a MD5 hash of the password.",
-        "InvalidNonceOrHeadersOrReferrer": "Form security failed. Please reload the form and check that your cookies are enabled. If you use a proxy server, you must %1$s configure Matomo to accept the proxy header%2$s that forwards the Host header. Also, check that your Referrer header is sent correctly.",
+        "InvalidNonceToken": "Form security failed, token miss match. Please reload the form and check that your cookies are enabled",
+        "InvalidNonceReferrer": "Form security failed, invalid referrer header, If you use a proxy server, you must %1$s configure Matomo to accept the proxy header%2$s that forwards the Host header. Also, check that your Referrer header is sent correctly and  If you previously connected using https, please ensure you are connecting over a secure (SSL/TLS) connection and try again.",
```
There are placeholders used but no value is actually passed to them in the translate method. When this error happens it looks like this:

![image](https://user-images.githubusercontent.com/273120/136727144-687cccf7-f2a3-4a33-bb27-4d7704a79f7b.png)
There should be also a dot (`.`) after `invalid referrer header` instead of a comma (`,`).

tsteur commented on plugins/Login/lang/en.json:
```diff
@@ -5,7 +5,11 @@
-        "InvalidNonceOrHeadersOrReferrer": "Form security failed. Please reload the form and check that your cookies are enabled. If you use a proxy server, you must %1$s configure Matomo to accept the proxy header%2$s that forwards the Host header. Also, check that your Referrer header is sent correctly.",
+        "InvalidNonceToken": "Form security failed, token miss match. Please reload the form and check that your cookies are enabled",
+        "InvalidNonceReferrer": "Form security failed, invalid referrer header, If you use a proxy server, you must %1$s configure Matomo to accept the proxy header%2$s that forwards the Host header. Also, check that your Referrer header is sent correctly and  If you previously connected using https, please ensure you are connecting over a secure (SSL/TLS) connection and try again.",
+        "InvalidNonceUnexpectedReferrer" : "Form security failed, the referrer header is different from the expected referrer header, check that your Referrer header is sent correctly.",
```
I wonder how someone would check if the referrer header is sent correctly. But let's ignore this for now as I suppose we can ignore it a bit for now as `$expectedReferrerHost` isn't used there so far. Nothing to do for now I reckon. Seems to be indeed not used in the code base.

tsteur commented on plugins/Login/lang/en.json:
```diff
@@ -5,7 +5,11 @@
+        "InvalidNonceToken": "Form security failed, token miss match. Please reload the form and check that your cookies are enabled",
+        "InvalidNonceReferrer": "Form security failed, invalid referrer header, If you use a proxy server, you must %1$s configure Matomo to accept the proxy header%2$s that forwards the Host header. Also, check that your Referrer header is sent correctly and  If you previously connected using https, please ensure you are connecting over a secure (SSL/TLS) connection and try again.",
+        "InvalidNonceUnexpectedReferrer" : "Form security failed, the referrer header is different from the expected referrer header, check that your Referrer header is sent correctly.",
+        "InvalidNonceOrigin": "Form security failed, invalid origin, If you previously connected using https, please ensure you are connecting over a secure (SSL/TLS) connection and try again",
```
Can we use a dot after `invalid origin` instead of a comma?

tsteur reviewed (changes requested):
@peterhashair left a comment as there's an issue with a translation.

Also the method `getMessageExceptionNoAccess` method in the login controller seems to be no longer used and we could remove it now?

The translation key `Login_InvalidNonceOrHeadersOrReferrer` seems to be still used in the `TwoFactorAuth/Controller`. Maybe in that controller we could similarly use the new method in the `loginTwoFactorAuth` to get the reason why the nonce check fails? Then we could remove the translation key `Login_InvalidNonceOrHeadersOrReferrer`.
![image](https://user-images.githubusercontent.com/273120/136728027-56370ae9-9e09-412d-bbfc-56e8aeeb6526.png)

Commit:
update nonce

remove getMessageExceptionNoAccess functions.  remove InvalidNonceOrHeadersOrReferrer from lang
move getMessageExceptionNoAccessWhenInsecureConnectionMayBeUsed to nonce check.

peterhashair commented:
- deprecated getMessageExceptionNoAccess Login/controller
- deprecated  InvalidNonceOrHeadersOrReferrer in lang
- move getMessageExceptionNoAccessWhenInsecureConnectionMayBeUsed from Login/controller to core/Nonce
- fix missing 2 variables in translation.

Commit:
update nonce additionalErrors

update nonce additionalErrors

tsteur commented on plugins/Login/lang/en.json:
```diff
@@ -5,7 +5,10 @@
         "ExceptionInvalidSuperUserAccessAuthenticationMethod": "A user with Super User access cannot be authenticated using the '%s' mechanism.",
         "ExceptionPasswordMD5HashExpected": "The password parameter is expected to be a MD5 hash of the password.",
-        "InvalidNonceOrHeadersOrReferrer": "Form security failed. Please reload the form and check that your cookies are enabled. If you use a proxy server, you must %1$s configure Matomo to accept the proxy header%2$s that forwards the Host header. Also, check that your Referrer header is sent correctly.",
+        "InvalidNonceToken": "Form security failed, token miss match. Please reload the form and check that your cookies are enabled",
```
here we need to add a dot at the end
![image](https://user-images.githubusercontent.com/273120/136888159-a7aa1e9c-8dc4-4cfb-9af2-6219d7fa09b6.png)

tsteur commented on plugins/Login/lang/en.json:
```diff
@@ -5,7 +5,10 @@
+        "InvalidNonceToken": "Form security failed, token miss match. Please reload the form and check that your cookies are enabled",
+        "InvalidNonceReferrer": "Form security failed, invalid referrer header. If you use a proxy server, you must %1$s configure Matomo to accept the proxy header %2$s that forwards the Host header. Also, check that your Referrer header is sent correctly and  If you previously connected using https, please ensure you are connecting over a secure (SSL/TLS) connection and try again.",
+        "InvalidNonceUnexpectedReferrer" : "Form security failed, the referrer header is different from the expected referrer header, check that your Referrer header is sent correctly.",
+        "InvalidNonceOrigin": "Form security failed, invalid origin. If you previously connected using https, please ensure you are connecting over a secure (SSL/TLS) connection and try again",
```
same here we need to add a dot here

tsteur commented on core/Nonce.php:
```diff
@@ -65,38 +65,80 @@ public static function getNonce($id, $ttl = 600)
      * @param string $id The nonce's unique ID. See {@link getNonce()}.
      * @param string $cnonce Nonce sent from client.
-     * @param string $expectedReferrerHost The expected referrer host for the HTTP referrer URL.
+     * @param null $expectedReferrerHost The expected referrer host for the HTTP referrer URL.
```
not too important, could change this to `null|string`

tsteur reviewed (changes requested):
@peterhashair last 2 comments to add a `dot` at the end of sentences. Then it should be good to merge.

Commit:
update phpdocs and lang file

update phpdocs and lang file
